### PR TITLE
Improve user agent handling/provide user agent "short" codes

### PR DESCRIPTION
**Pull request by TheCleric**
So I took a risk with this PR (and probably should have discussed some of this with you first @ronjouch, but I started tugging on the thread of:

**How can I get Google (and other apps) to behave better or make it easier for a user to be self-sufficient?**

Obviously this has been an ongoing problem, but particularly in this thread, and this comment: https://github.com/nativefier/nativefier/issues/1193#issuecomment-842266213 . Not that there was anything WRONG in this question, it just got me thinking that there has to be a better way.

So this is my attempt at a "better" way. Better being in the eye of the developer though, I'm fine changing some of this until we're both happy.

Here's what I've changed:
- Eliminated the "inferUserAgent" logic, because I realized we can:
- Rely more heavily on Electron's `app.userAgentFallback` which has a Chrome-like user agent pre-populated in it already, based on the Electron version used. For `--honest`/`--user-agent-honest` we'll use it as is. Without that we'll strip out the elements of it that indicate it's not Chrome (Electron/version and GeneratedAppName/version)
- Add short codes that can be passed to `--user-agent` that will determine the latest versions of some common browsers, and generate a user agent for the user, so they don't have to go hunting one down. Included is support for:
    -  `edge` which will grab the chrome version by electron version like `inferUserAgent` did, and populate it
    - `firefox` which will hit the FF versions API and get latest stable and generate a user agent
    - `safari` which will hit Wikipedia (sorry, it's the only reliable, up to date source I could find) and get the latest version (and WebKit version) and generate a user agent
- Standardize on using `userAgentHonest` over `honest`

### CODE REVIEW

File: app/src/main.ts [hunk 1]
```diff
@@ -46,6 +46,17 @@ if (appArgs.portable) {
...
 
+if (!appArgs.userAgentHonest) {
+  // Electron userAgentFallback is the user agent used if none is specified when creating a window.
+  // For our purposes, it's useful because its format is similar enough to a real Chrome's user agent to not need
+  // to infer the userAgent. userAgentFallback normally looks like this:
+  // Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) app-nativefier-804458/1.0.0 Chrome/89.0.4389.128 Electron/12.0.7 Safari/537.36
+  // We just need to strip out the appName/1.0.0 and Electron/electronVersion
+  app.userAgentFallback = app.userAgentFallback
+    .replace(`Electron/${process.versions.electron}`, '')
+    .replace(`${app.getName()}/${app.getVersion()}`, ' ');
```
Comment (ronjouch) [inline app/src/main.ts:57]:
```suggestion
    .replace(`Electron/${process.versions.electron} `, '')
    .replace(`${app.getName()}/${app.getVersion()} `, ' ');
```
? Else we'd get weird double spaces, don't we? Is this unit-tested?

File: src/infer/browsers/inferFirefoxVersion.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+export async function getLatestFirefoxVersion(
+  url = FIREFOX_VERSIONS_URL,
+): Promise<string> {
+  log.debug('Grabbing apple version data from', url);
+  const response = await axios.get(url, { timeout: 5000 });
+  if (response.status !== 200) {
+    throw new Error(`Bad request: Status code ${response.status}`);
+  }
+  const firefoxVersions: FirefoxVersions = response.data;
+
```
Comment (ronjouch) [inline src/infer/browsers/inferFirefoxVersion.ts:34]:
```suggestion

  log.debug(`Got latest Firefox version ${firefoxVersions.LATEST_FIREFOX_VERSION}`);
```

File: src/infer/browsers/inferFirefoxVersion.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  'https://product-details.mozilla.org/1.0/firefox_versions.json';
+
+export async function getLatestFirefoxVersion(
+  url = FIREFOX_VERSIONS_URL,
+): Promise<string> {
+  log.debug('Grabbing apple version data from', url);
+  const response = await axios.get(url, { timeout: 5000 });
+  if (response.status !== 200) {
+    throw new Error(`Bad request: Status code ${response.status}`);
+  }
```
Comment (ronjouch) [inline src/infer/browsers/inferFirefoxVersion.ts:32]: Error handling: can you wrap this in a try/catch and return a default version that is fetched from a new `DEFAULT_FIREFOX_VERSION` defined in `const.ts` (visible next to `DEFAULT_ELECTRON_VERSION` , with a comment reminding to periodically bump this to whatever's the latest Firefox) ?

File: src/infer/browsers/inferFirefoxVersion.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  NEXT_SOFTFREEZE_DATE: string;
+};
+
+const FIREFOX_VERSIONS_URL =
+  'https://product-details.mozilla.org/1.0/firefox_versions.json';
+
+export async function getLatestFirefoxVersion(
+  url = FIREFOX_VERSIONS_URL,
+): Promise<string> {
+  log.debug('Grabbing apple version data from', url);
```
Comment (ronjouch) [inline src/infer/browsers/inferFirefoxVersion.ts:28]:
```suggestion
  log.debug('Grabbing Firefox version data from', url);
```

File: src/infer/browsers/inferSafariVersion.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        ...versionRow.matchAll(/>\s*(([0-9]*\.){3,4}[0-9])\s*</g),
+      ];
+      if (webkitVersionMatch.length > 0 && !webkitVersion) {
+        webkitVersion = webkitVersionMatch[0][1];
+      }
+      if (version && webkitVersion) {
+        break;
+      }
+    }
+
```
Comment (ronjouch) [inline src/infer/browsers/inferSafariVersion.ts:62]: Eeek :grimacing: .

What about simply having a `DEFAULT_SAFARI_VERSION` in const.ts, with a link to the Wikipedia page, and a comment reminding to bump it when we bump Electron?

Alternatively, if you insist on keeping this code, we can still have the fallback with `DEFAULT_SAFARI_VERSION` frmo const.ts in the `catch` below (as I suggested for Firefox).

File: docs/api.md [hunk 6]
```diff
@@ -52,10 +52,13 @@
     - [[zoom]](#zoom)
```
Comment (ronjouch) [inline docs/api.md:52]: Can you also update `CATALOG.md`, and replace `--user-agent <user agent of the latest Firefox>` with `--user-agent firefox` ?

**Review by ronjouch — APPROVED** [review]
Looks good, I have a couple nits.

> "Add short codes [so that users] don't have to go hunting one down"

Great idea 👍

File: app/src/main.ts (hunk 1 above)
Comment (TheCleric) [reply app/src/main.ts:57]: > ? Else we'd get weird double spaces, don't we? Is this unit-tested?

It's not because we don't have any unit tests for this file right now. I'll move it to a helper function and add some tests.

File: src/infer/browsers/inferFirefoxVersion.ts (hunk 4 above)
Comment (TheCleric) [reply src/infer/browsers/inferFirefoxVersion.ts:28]: Whoops!

File: src/infer/browsers/inferSafariVersion.ts (hunk 5 above)
Comment (TheCleric) [reply src/infer/browsers/inferSafariVersion.ts:62]: > Eeek 😬 .

Yeah. 😭 

> What about simply having a `DEFAULT_SAFARI_VERSION` in const.ts, with a link to the Wikipedia page, and a comment reminding to bump it when we bump Electron?
> 
> Alternatively, if you insist on keeping this code, we can still have the fallback with `DEFAULT_SAFARI_VERSION` frmo const.ts in the `catch` below (as I suggested for Firefox).

I insist on nothing here. 😆 

Here's what it comes down to for me, and I'll let you make the call. My only goal here is to limit the pain of keeping something up to date and generating user agents. If this code keeps us from having to keep another constant up to date, good. But if this code is so brittle (and it could be depending on the whims of a Wikipedia editor) that it breaks and then we have to support THAT, then it's not good.

So basically we just have to make a bet on which one is easier to maintain (so probably just a constant, but I'll let you make the call).

Also one thing I was trying to keep in mind here is that I don't want someone looking to block Nativefier to be able to easily guess what the user agent would be here. But that may be paranoia. 🤣

Comment (ronjouch) [reply src/infer/browsers/inferSafariVersion.ts:62]: @TheCleric

1. Okay to give it a try, we'll see how long it works...
2. ... with fallback to a constant in the `catch`
2. ... and with more logging leaving a trace of what happened, so that it's auditable by users. Something like "Trying to get Safari versions from the web... \n Failed to get Safari version from the web, falling back to $var_from_const_dot_ts"

**Comment by TheCleric** [comment]
Alright @ronjouch I think all nits have been addressed

**Review by ronjouch — APPROVED** [review]
👍:clap:
